kroxylicious/src/main/java/io/kroxylicious/proxy/internal/filter/EagerMetadataLearner.java

```diff
@@ -0,0 +1,83 @@
+ */
+package io.kroxylicious.proxy.internal.filter;
+
+import java.util.Set;
+
+import org.apache.kafka.common.message.MetadataRequestData;
+import org.apache.kafka.common.message.MetadataResponseData;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.filter.KrpcFilterContext;
+import io.kroxylicious.proxy.filter.RequestFilter;
+
+/**
+ * An internal filter that causes the system to eagerly learn the cluster's topology by spontaneously emitting
+ * an out-of-band Metadata request at the earliest legal point in the Kafka conversation.  The response to allows
+ * the Endpoint reconciliation to take place so that ephemeral upstream bindings are replaced by true bindings to
+ * the actual upstream brokers.
+ * <br/>
+ * Once the bindings are made, the filter causes the client's connection to close.   This is done
+ * in order to force the client to reconnect, thus ensuring the client has a connection to the intended broker.
+ *
+ * @see io.kroxylicious.proxy.internal.net.EndpointRegistry
+ */
+public class EagerMetadataLearner implements RequestFilter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EagerMetadataLearner.class);
+
+    /**
+     * The set of the API keys that are permitted before the client would normally send a METADATA request.
+     */
+    private final static Set<ApiKeys> KAFKA_PRELUDE = Set.of(ApiKeys.API_VERSIONS, ApiKeys.SASL_HANDSHAKE, ApiKeys.SASL_AUTHENTICATE);
+
+    public EagerMetadataLearner() {
+    }
+
+    @Override
+    public void onRequest(ApiKeys apiKey, RequestHeaderData header, ApiMessage body, KrpcFilterContext filterContext) {
+        if (KAFKA_PRELUDE.contains(apiKey)) {
+            filterContext.forwardRequest(header, body);
+        }
+        else {
+            final short apiVersion = determineMetadataApiVersion(header);
+            // Send an out-of-band Metadata request. The response will be intercepted by the in-built BrokerAddressFilter.
+            // By the time control returns to the handler, the upstream addresses will have been reconciled.
+            var unused = filterContext.sendRequest(apiVersion, new MetadataRequestData()).thenAccept(apiMessage -> {
+                if (!(apiMessage instanceof MetadataResponseData metadataResponseData)) {
+                    throw new IllegalStateException("Unexpected response " + apiMessage.getClass() + " to out-of-band request.");
+                }
+                if (apiKey.equals(ApiKeys.METADATA) && apiVersion == header.requestApiVersion()) {
+                    // The client's requested matched our out-of-band request, so we may as well return the
+                    // response.
+                    filterContext.forwardResponse(metadataResponseData);
+                }
```

Comment:
My reasoning was that it was better to give the client the information as soon as we could.   Since the request was for metadata, we might as well give it to them.  I think there will be scenarios where the client is saved a round trip to the server by behaving in this way.

I don't think there's an out of order concern.

Reply:
Yeah since everything is executed on the same eventloop I think we know that this message should be sent back to the client in order. The ordering problem happens if we hand off work to other threads and the netty loop is free to forward more messages.